Sample content

calherries opened the pull request:
Epic: https://github.com/metabase/metabase/issues/40066

Follows https://github.com/metabase/metabase/pull/40907, which was a pre-requisite for this change

To verify:
- drop the app DB, and if using Postgres or MySQL re-create it
- start up metabase

During setup, we insert sample content from `resources/sample-content.edn` into the app DB as a migration.
The sample content includes the "Examples" collection, which contains the "E-commerce insights" dashboard and its dependencies.

The approach avoids having to maintain any code to generate the collection and its contents, because it's just data. This approach ensures the contents is forward compatible to the same extent that any dashboard/card/collection is forward compatible, because we have to keep these entities forward compatible anyway.

The content comes from this collection: https://stats.metabase.com/collection/1449-example-dashboard-prototype

### Testing strategy

The testing strategy for this is controversial, and worth questioning. There's two things you should know:

1. e2e tests run without the sample content
There are two reasons for this: 
(a) we don't want to couple e2e tests to the sample content to make it easy to update
(b) updating the e2e tests are a pain, because there are a lot of tests that depend on the sample content not being there

However, I would argue we can get away with minimal test coverage using the following logic:

```
1. we promise users that "previously created content" will work after any upgrade in the future
2. we (should) have tests that enforce this property
3. whenever an upgrade happens, the sample content is always "previously created content" because it is created when a metabase app DB is initialized as a DB migration
4. therefore, we have tests to enforce that the sample content will work after any upgrade in the future
```

Also
```
1. we have tests that check the behaviour of content created with the metabase UI works as expected
2. the sample content is created with the metabase UI
3. therefore, we have tests that prove the sample content works as expected
```
2. Most of the backend tests run without the sample content. This is just to keep the test suite fast, because with the sample content the postgres and mysql test suites were taking over 90 minutes to complete. This carries substantial risk, but I don't have an alternative solution. There's potentially room to increase the number of tests where the sample content is present and it could be important though, e.g. loading from h2 tests. My worry is any code that depends on an initialized metabase instance having no content may pass tests but fail in production. As far as I'm aware there's no existing code of this nature but it's possible this could exist in the future. At least in development, we would catch the issue.

calherries committed:
initial dashboard loading

calherries committed:
use the example dashboard

calherries committed:
Update password of user

calherries committed:
Depend on `MB_LOAD_SAMPLE_CONTENT` environment variable

calherries committed:
Update comment

calherries committed:
Use auto-generated IDs

calherries committed:
Fix comment

calherries committed:
fix comment

calherries committed:
Add tests

calherries committed:
Fix test

calherries committed:
Fix copying from h2

calherries committed:
Fix test

npretto commented on resources/sample-content.edn:
```diff
@@ -0,0 +1,4236 @@
+   :updated_at #t "2024-04-01T15:03:43.314320Z",
+   :cache_ttl nil,
+   :details
+   "{\"db\":\"file:/Users/callumherries/meta/metabase/plugins/sample-database.db;USER=GUEST;PASSWORD=guest\"}",
```
Should this be here?

calherries replied:
Ha. thanks for the reminder, I'll delete this in the final version :D The `details` get updated later to reflect the real location of the sample DB, but I'll delete this manually anyway. gotta keep the JAR size down

calherries committed:
Remove sample content in tests

calherries committed:
Fix

calherries committed:
Create the internal user in a migration

calherries committed:
Add internal user tests

calherries committed:
Remove internal user being created for EE

calherries committed:
Don't insert the core_user again

calherries committed:
Remove setup test

calherries committed:
Use collection as a proxy to abort migration

calherries committed:
Merge branch 'master' into cal-sample-dashboard

calherries committed:
Don't load sample content in tests :(

calherries committed:
fix test

calherries committed:
restore sample database creation if it doesn't exist

calherries committed:
Skip creating sample content selectively

calherries committed:
fix test

calherries committed:
don't load sample content in e2e tests

calherries committed:
Fix sample content

calherries committed:
Add tests

calherries committed:
skip again

calherries committed:
Remove my details

calherries committed:
Actually skip e2e tests

calherries committed:
Remove comment

crisptrutski commented on test/metabase/test/data.clj:
```diff
@@ -269,5 +269,5 @@
      (with-redefs [perms-group/all-users (#'perms-group/magic-group perms-group/all-users-group-name)
                    perms-group/admin     (#'perms-group/magic-group perms-group/admin-group-name)]
-       (mdb/setup-db!)
+       (mdb/setup-db! :create-sample-content? false) ; skip sample content for speedy tests. this doesn't reflect production
```
Oh, I see you do repeat this for a bunch of sites - perhaps just add the relevant comment (for speed or to avoid brittleness) to the non-annotated sites too.

I also feel that this should be a line comment above.

calherries replied:
I should have added this comment to every site. Did you see a site where this comment is missing?

> this should be a line comment

I added the comment on the end of the line to grep easier to check all sites have the comment. don't care enough to change it

crisptrutski replied:
Fair. I wish we had a "structural" find-replace + a reliable auto formatter.

crisptrutski replied:
> Did you see a site where this comment is missing?

For example `cmd.clj` and [this test](https://github.com/metabase/metabase/pull/40753/files#diff-f039ac289d7def5fe424d733725acfdc2cb359516bad4cbdcf2d0078ae2b9a94R34) - I think it's worth going whole hog.

calherries replied:
Ah thanks for the test, that test should actually have `true` because that one test isn't going to kill test speed. 

For `cmd.clj`, I thought it would be obvious why we don't want the sample content loaded though, based on the context. The purpose of skipping it is not for speed.

crisptrutski reviewed (commented):
LGTM to me 

> e2e tests will run without the sample content

Motivation and defense seem totally reasonable.

> Most of the backend tests run without the sample content

I feel that I'm missing something here re: substantial risk. Besides the "load-from-h2" scenario, what's the worst kind of issue you could anticipate slipping through?

calherries committed:
Remove :maybe

calherries commented on src/metabase/cmd/copy.clj:
```diff
@@ -382,15 +383,15 @@
   ;; make sure the source database is up-do-date
   (step (trs "Set up {0} source database and run migrations..." (name source-db-type))
-    (mdb.setup/setup-db! source-db-type source-data-source true))
+    (mdb.setup/setup-db! source-db-type source-data-source true false))
```
Ha, I was thinking you might catch that. The only reason I didn't do it because I was lazy, but since you like keyword arguments (as I do) I will follow through with it

crisptrutski replied:
I should also have said - fine for expedience, if you'll fix in a later PR. I suspected as much.

calherries replied:
yep, will do in a follow up PR

calherries commented on src/metabase/db/setup.clj:
```diff
@@ -158,13 +158,15 @@
 (mu/defn setup-db!
   "Connects to db and runs migrations. Don't use this directly, unless you know what you're doing;
   use [[metabase.db/setup-db!]] instead, which can be called more than once without issue and is thread-safe."
```
Yeah, deleted

calherries commented on src/metabase/db.clj:
```diff
@@ -53,8 +53,10 @@
-  database migrations. If DB is already set up, this function will no-op. Thread-safe."
-  []
+  database migrations. If DB is already set up, this function will no-op. Thread-safe.
+  Callers must decide whether or not to create sample content during migrations."
```
The value is arguably minimal but I wanted it to be an explicit decision for every caller because in many cases where it's false, it's because we're skipping it to make tests faster. I want to highlight the danger in those cases. And I don't want to make it true by default because there are cases where that is obviously wrong e.g. load-from-h2 or importing serialized content.

crisptrutski replied:
I've come around, I like every site having an explicit choice coupled with an explicit explanation comment.

calherries commented on enterprise/backend/test/metabase_enterprise/serialization/test_util.clj:
```diff
@@ -72,7 +72,7 @@
     ;; DB should stay open as long as `conn` is held open.
     (with-open [_conn (.getConnection data-source)]
-      (with-db data-source (mdb/setup-db!))
+      (with-db data-source (mdb/setup-db! :create-sample-content? false)) ; skip sample content for speedy tests. this doesn't reflect production
```
> I feel that this comment belongs in every test

I'm not sure I understand. I have added this comment in every case where `:create-sample-content?=false` in tests for this reason. maybe you mean something more than that?

Yeah good idea about adding it to the docstring, I will do that

crisptrutski replied:
This comment came from other tests disabling the sample content without any comment, e.g. the e2e tests. Since this was the only test with any comment until much further down in the diff, I thought it might be the only spot. I think it's worth noting the relevant reason for every test that explicitly disables it.

crisptrutski replied:
Basically a repeat of my later comment https://github.com/metabase/metabase/pull/40753#discussion_r1551276288

calherries replied:
what tests are missing the comment? I thought I got everything

calherries replied:
resolving this because we switched to other comment thread. On the docstring, I have added it in 8dc50dd90ef9683979ed4d047d88992c8720c826

calherries commented on src/metabase/core.clj:
```diff
@@ -110,7 +110,11 @@
+  ;; we have to skip creating sample content if we're running tests, because it causes some tests to timeout
+  ;; and the test suite can take 2x longer. this is really unfortunate because it could lead to some false
+  ;; negatives, but for now there's not much we can do
+  (mdb/setup-db! :create-sample-content? (not config/is-test?))
```
Fine idea, I'll bring it up at the guild meeting today.

calherries committed:
tidy comments per chris' lovely suggestions

calherries committed:
flip to true

calherries commented:
> I feel that I'm missing something here re: substantial risk. Besides the "load-from-h2" scenario, what's the worst kind of issue you could anticipate slipping through?

This is a very good question, and I don't have a great answer. I don't know what risks there are, so maybe you could characterise it as "uncertainty". But any code that depends on an initialized metabase instance having no content may pass tests but fail in production. As far as I'm aware there's no existing code of this nature but it's possible this could exist in the future

albertoperdomo reviewed (changes requested):
* It looks like we could tweak a few of the question names - @vbenedetti ?
  * Some questions have `- Maz` in the name. We should probably drop that
  * (not binned yet) - is this question used or was it replaced with a new one
  * Maybe we should not use the auto-generated names? 
  
 Also, is it me? But I can't seem to find the dashboard itself?

calherries committed:
Update sample content (#41201)

calherries committed:
Merge branch 'master' into cal-sample-dashboard

calherries commented:
@albertoperdomo @vbenedetti I've tweaked the content to incorporate these changes:
- everything is one collection instead of two
- changes in titles and dashboards [notion.so/metabase/Example-dashboard-final-copy-c0c6b2b8eb114d5b832eb5612a8653a1?pvs=6&utm_content=c0c6b2b8-eb11-4d5b-832e-b5612a8653a1&utm_campaign=T078VCLCR&n=slack&n=slack_link_unfurl](https://www.notion.so/metabase/Example-dashboard-final-copy-c0c6b2b8eb114d5b832eb5612a8653a1?pvs=6&utm_content=c0c6b2b8-eb11-4d5b-832e-b5612a8653a1&utm_campaign=T078VCLCR&n=slack&n=slack_link_unfurl)

calherries committed:
Fix merge conflict

calherries committed:
Fix merge conflict

calherries committed:
Fix sample content IDs

calherries committed:
Update content

calherries committed:
Update filters and graphs

calherries committed:
fix migration and delete card

calherries committed:
Merge branch 'master' into cal-sample-dashboard

calherries commented:
@albertoperdomo @vbenedetti I combed through the content and realised there were a few errors in the queries and places where the dashboard could be improved. I've made the following changes:
- "Buyer's ages by segments" (which I renamed to "Buyers by age group") was showing sum of counts of people by individual ages, but the x-axis was the count of people in each age, not the age.
  Now it shows the number of people in each 5 year age range.
- There were a few questions in "Demographics" that had auto-binned ages, which meant strange bins of 7.5 years, e.g. "22.5-30". Now they have bins of 5 years, e.g. "20-25".
- Trend charts like "Revenue per quarter" had N/A and now they report the last two months.
- Time-series graphs had either hard-coded dates or no date filters applied at all.
  Now I added filters for these questions with orders.created_at in the last 24 months
- "Unique customers per month" was using people.created_at, which grouped customers by the time they signed up. So I changed it to group by orders.created_at
- The Location filter (which searches people.state) was an input box, so I changed it to a dropdown list

I'd appreciate if you (or one of you) double-checked these changes, comparing against [stats](https://stats.metabase.com/dashboard/2190-e-commerce-insights), approve this PR when you're done.

calherries committed:
Changesets in order

calherries committed:
Update migration ID in test

calherries committed:
Make dashboard full width

calherries committed:
Create example-dashboard-id setting (#41272)
